### PR TITLE
[MU4] KDAB dockwidgets integration. Docks

### PR DESCRIPTION
**Pull request by RomanPudashkin**
- added all types of docks based on KDAB dockwidgets

### CODE REVIEW

File: src/appshell/view/dockwindow_kdab/dockpage.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    : QQuickItem(parent),
+    m_toolBars(this),
+    m_panels(this),
+    m_statusBars(this)
+{
+}
+
+QString DockPage::uniqueName() const
+{
+    return m_uniqueName;
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/dockpage.cpp:45]: I think it better to use `objectName`

File: src/appshell/view/dockwindow_kdab/dockpage.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    window.addDock(m_central, KDDockWidgets::Location_OnRight);
+
+    for (DockPanel* panel : m_panels.list()) {
+        window.addDock(panel, KDDockWidgets::Location_OnLeft);
+    }
+
+    DockToolBar* prevToolBar = nullptr;
+    for (DockToolBar* toolBar : m_toolBars.list()) {
+        auto location = prevToolBar ? KDDockWidgets::Location_OnRight : KDDockWidgets::Location_OnTop;
+        window.addDock(toolBar, location, prevToolBar);
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/dockpage.cpp:116]: Why is there such a location? I believe we need an explanation (comment)
And it seems that location should be a property of the toolbar

File: src/appshell/view/dockwindow_kdab/dockpage.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        prevToolBar = toolBar;
+    }
+
+    DockStatusBar* prevStatusBar = nullptr;
+    for (DockStatusBar* statusBar : m_statusBars.list()) {
+        auto location = prevStatusBar ? KDDockWidgets::Location_OnRight : KDDockWidgets::Location_OnBottom;
+        window.addDock(statusBar, location, prevStatusBar);
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/dockpage.cpp:123]: Why is there such a location? I believe we need an explanation (comment)
And it seems that location should be a property of the toolbar

File: src/appshell/view/dockwindow_kdab/docktoolbar.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+Qt::Orientation DockToolBar::orientation() const
+{
+    return m_orientation;
+}
+
+QStringList DockToolBar::allowedPagesUriList() const
+{
+    return m_allowedPagesUriList;
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/docktoolbar.cpp:51]: What is it for?

File: src/appshell/view/dockwindow_kdab/dockwindow.cpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    newPage->init(*this);
+
+    DockToolBar* prevToolBar = nullptr;
+    for (DockToolBar* toolBar : m_toolBars.list()) {
+        auto location = prevToolBar ? KDDockWidgets::Location_OnRight : KDDockWidgets::Location_OnTop;
+        addDock(toolBar, location, prevToolBar);
+
+        if (!toolBar->allowedPagesUriList().empty() && !toolBar->allowedPagesUriList().contains(uri)) {
+            toolBar->close();
+        }
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/dockwindow.cpp:107]: Confusing.
As far as I can see from the code (and as it was before), we have:
1. Toolbars of the window - always shown
2. Page toolbars - shown if the page is shown

It is strange to see the window toolbar that needs to be hidden for the given pages. Maybe then such a toolbar can simply be move to the page?

File: src/appshell/view/dockwindow_kdab/internal/dockbase.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+QString DockBase::title() const
+{
+    return m_title;
+}
+
+QString DockBase::uniqueName() const
+{
+    return m_uniqueName;
+}
```
Comment (igorkorsukov) [inline src/appshell/view/dockwindow_kdab/internal/dockbase.cpp:46]: Better to use `objectName`

File: src/appshell/view/dockwindow_kdab/dockpage.cpp (hunk 1 above)
Comment (RomanPudashkin) [reply src/appshell/view/dockwindow_kdab/dockpage.cpp]: fixed

File: src/appshell/view/dockwindow_kdab/dockpage.cpp (hunk 2 above)
Comment (RomanPudashkin) [reply src/appshell/view/dockwindow_kdab/dockpage.cpp]: added a comment

File: src/appshell/view/dockwindow_kdab/docktoolbar.cpp (hunk 4 above)
Comment (RomanPudashkin) [reply src/appshell/view/dockwindow_kdab/docktoolbar.cpp]: removed

File: src/appshell/view/dockwindow_kdab/dockwindow.cpp (hunk 5 above)
Comment (RomanPudashkin) [reply src/appshell/view/dockwindow_kdab/dockwindow.cpp]: fixed

**Review by igorkorsukov — APPROVED** [review]

**Pull request merged**